File: library-java/lib/src/main/java/eu/postyard/templates/RenderedTemplate.java
package eu.postyard.templates;

import java.util.Collections;
import java.util.Map;

/**
 * The result of rendering a {@link Template}: the rendered output of each block, keyed by block
 * name.
 */
public final class RenderedTemplate {

  private final Map<String, String> blocks;

  /**
   * Wraps the rendered output of each block.
   *
   * @param blocks the rendered output of each block, keyed by block name
   */
  public RenderedTemplate(Map<String, String> blocks) {
    this.blocks = Collections.unmodifiableMap(blocks);
  }

  /**
   * Returns the rendered output of a single block.
   *
   * @param blockName the block's name
   * @return the rendered output, or {@code null} if the template has no such block
   */
  public String get(String blockName) {
    return blocks.get(blockName);
  }

  /** {@return an unmodifiable view of every block's rendered output, keyed by name} */
  public Map<String, String> blocks() {
    return blocks;
  }
}

File: library-typescript/src/RenderedTemplate.ts
/**
 * The output of rendering a {@link Template}: each block's rendered string,
 * keyed by block name.
 */
export class RenderedTemplate {
  private readonly blocks: ReadonlyMap<string, string>;

  /**
   * @param blocks - The rendered output of each block, keyed by block name.
   */
  constructor(blocks: Map<string, string>) {
    this.blocks = new Map(blocks);
  }

  /**
   * Returns the rendered output of a single block.
   *
   * @param blockName - The block's name.
   * @returns The block's rendered output.
   * @throws If the template has no block with that name.
   */
  get(blockName: string): string {
    const value = this.blocks.get(blockName);
    if (value === undefined) {
      throw new Error(`Block '${blockName}' not found`);
    }
    return value;
  }

  /**
   * @returns The names of the rendered blocks, in render order.
   */
  blockNames(): string[] {
    return [...this.blocks.keys()];
  }
}
